Stack Trace Parsing Stack Healthcheck

Some `curl` based healthchecks for stacks are not parsed correctly resulting in a stack trace.

A user reported a stack trace when deploying from a docker-compose.yml file.
The trace indicated that we were out of bounds on a slice reference when parsing the curl health check for the service.

I was able to repro it with the following unittest

```
func TestTranslateHealthcheckCurlToHttp(t *testing.T) {
	check := HealthCheck{
		Test:        []string{"curl http://0.0.0.0:4572"},
		Interval:    5 * time.Second,
		Timeout:     10 * time.Minute,
		Retries:     10,
		StartPeriod: 20 * time.Second,
	}

	translateHealtcheckCurlToHTTP(&check)
	fmt.Printf("CHECK %+v\n", check)
}
```

This test results in a stack of

```
--- FAIL: TestTranslateHealthcheckCurlToHttp (0.00s)
panic: runtime error: slice bounds out of range [20:11] [recovered]
        panic: runtime error: slice bounds out of range [20:11]

goroutine 36 [running]:
testing.tRunner.func1.2({0x19148a0, 0xc000164a80})
        /usr/local/go/src/testing/testing.go:1209 +0x24e
testing.tRunner.func1()
        /usr/local/go/src/testing/testing.go:1212 +0x218
panic({0x19148a0, 0xc000164a80})
        /usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/okteto/okteto/pkg/model.translateHealtcheckCurlToHTTP(0xc000080ed0)
        /Users/jdm/okt/okteto/pkg/model/stack_serializer.go:520 +0x365
github.com/okteto/okteto/pkg/model.TestTranslateHealthcheckCurlToHttp(0x0)
        /Users/jdm/okt/okteto/pkg/model/stack_serializer_test.go:39 +0xc5
testing.tRunner(0xc0001e3a00, 0x19e7c98)
        /usr/local/go/src/testing/testing.go:1259 +0x102
created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:1306 +0x35a
FAIL    github.com/okteto/okteto/pkg/model      0.253s
FAIL
```

The user was on okteto CLI 1.14.5

Their check did not repro for me and was like this but it seems to be similar to the test based on the stack

```
    healthcheck:
      test: [ "CMD-SHELL", "curl http://0.0.0.0:4578" ]
      interval: 5s
      timeout: 10m
      retries: 10
      start_period: 20s
```

The `translateHealtcheckCurlToHTTP` function does a lot of manual parsing. We should investigate using `url.Parse` and other methods for doing some of the parsing while addressing the issue.

--- a/pkg/model/stack_serializer.go
+++ b/pkg/model/stack_serializer.go
@@ -490,7 +490,6 @@ func validateHealthcheck(healthcheck *HealthCheck) error {
 }
 
 func translateHealtcheckCurlToHTTP(healthcheck *HealthCheck) {
-	fmt.Printf("HC %+v\n", healthcheck)
 	// Join and then split the strings by space to ensure that
 	// each element in the string slice is a contiguous string with
 	// no spaces.
@@ -527,14 +526,11 @@ func translateHealtcheckCurlToHTTP(healthcheck *HealthCheck) {
 	}
 
 	if checkURL == nil {
-		fmt.Println("NOT FOUND")
 		return
 	}
 
-	fmt.Printf("URL FOUND: %+v\n", checkURL)
 	p := checkURL.Port()
 	if p == "" {
-		fmt.Println("PORT")
 		return
 	}
 	port, err := strconv.Atoi(p)
@@ -551,7 +547,6 @@ func translateHealtcheckCurlToHTTP(healthcheck *HealthCheck) {
 
 	healthcheck.HTTP = &HTTPHealtcheck{Path: path, Port: int32(port)}
 	healthcheck.Test = make(HealtcheckTest, 0)
-	fmt.Printf("NEXT %+v\n", healthcheck)
 }
 
 func getSvcPorts(public bool, rawPorts, rawExpose []PortRaw) (bool, []Port, error) {

--- a/pkg/model/stack_serializer.go
+++ b/pkg/model/stack_serializer.go
@@ -15,7 +15,7 @@ package model
 
 import (
 	"fmt"
-	"regexp"
+	"net/url"
 	"strconv"
 	"strings"
 	"time"
@@ -490,45 +490,68 @@ func validateHealthcheck(healthcheck *HealthCheck) error {
 }
 
 func translateHealtcheckCurlToHTTP(healthcheck *HealthCheck) {
-	localPortTestRegex := `^curl ((-f|--fail) )?'?((http|https)://)?(localhost|0.0.0.0):\d+(\/\w*)?'?$`
-	regexp, err := regexp.Compile(localPortTestRegex)
+	fmt.Printf("HC %+v\n", healthcheck)
+	// Join and then split the strings by space to ensure that
+	// each element in the string slice is a contiguous string with
+	// no spaces.
+	s := strings.Join(healthcheck.Test, " ")
+	testStrings := strings.Split(s, " ")
+
+	// There should be at least two strings, the curl binary and url.
+	if len(testStrings) < 2 {
+		return
+	}
+
+	if testStrings[0] != "curl" {
+		return
+	}
+
+	var checkURL *url.URL
+	for _, possibleURL := range testStrings[1:] {
+		if possibleURL == "-f" || possibleURL == "--fail" {
+			continue
+		}
+
+		u, err := url.ParseRequestURI(possibleURL)
+		// It's possible to have a healthcheck url without the scheme. If
+		// that happens then we inject one and attempt parsing again.
+		if err != nil || u.Host == "" {
+			u, err = url.ParseRequestURI("https://" + possibleURL)
+			if err != nil {
+				u = nil
+				continue
+			}
+		}
+		checkURL = u
+		break
+	}
+
+	if checkURL == nil {
+		fmt.Println("NOT FOUND")
+		return
+	}
+
+	fmt.Printf("URL FOUND: %+v\n", checkURL)
+	p := checkURL.Port()
+	if p == "" {
+		fmt.Println("PORT")
+		return
+	}
+	port, err := strconv.Atoi(p)
 	if err != nil {
 		return
 	}
-	testString := strings.Join(healthcheck.Test, " ")
-	if regexp.MatchString(testString) {
-		var firstSlashIndex, portStart int
-		if strings.Contains(testString, "://") {
-			testStringCopy := testString
-			for i := 0; i < 3; i++ {
-				firstSlashIndex += strings.Index(testStringCopy[firstSlashIndex:], "/") + 1
-			}
-			testStringCopy = testString
-			for i := 0; i < 2; i++ {
-				portStart += strings.Index(testStringCopy[portStart:], ":") + 1
-			}
-			portStart--
-			firstSlashIndex--
-		} else {
-			firstSlashIndex = strings.Index(testString, "/")
-			portStart = strings.Index(testString, ":")
-		}
 
-		var port, path string
-		if firstSlashIndex != -1 {
-			port = testString[portStart+1 : firstSlashIndex]
-			path = testString[firstSlashIndex:]
-		} else {
-			port = testString[portStart+1:]
-			path = "/"
-		}
-		p, err := strconv.Atoi(port)
-		if err != nil {
-			return
-		}
-		healthcheck.HTTP = &HTTPHealtcheck{Path: path, Port: int32(p)}
-		healthcheck.Test = make(HealtcheckTest, 0)
+	var path string
+	if checkURL.Path == "" {
+		path = "/"
+	} else {
+		path = checkURL.Path
 	}
+
+	healthcheck.HTTP = &HTTPHealtcheck{Path: path, Port: int32(port)}
+	healthcheck.Test = make(HealtcheckTest, 0)
+	fmt.Printf("NEXT %+v\n", healthcheck)
 }
 
 func getSvcPorts(public bool, rawPorts, rawExpose []PortRaw) (bool, []Port, error) {

--- a/pkg/model/stack_serializer_test.go
+++ b/pkg/model/stack_serializer_test.go
@@ -14,6 +14,7 @@
 package model
 
 import (
+	"fmt"
 	"os"
 	"reflect"
 	"testing"
@@ -25,6 +26,77 @@ import (
 	"k8s.io/apimachinery/pkg/api/resource"
 	"k8s.io/utils/pointer"
 )
+
+func TestTranslateHealthcheckCurlToHttp(t *testing.T) {
+	for _, tc := range []struct {
+		name         string
+		test         []string
+		expectedTest HealtcheckTest
+		expectedHTTP HTTPHealtcheck
+	}{
+		{
+			name:         "RegularCurlWithZeros",
+			test:         []string{"curl http://0.0.0.0:4572"},
+			expectedTest: HealtcheckTest{},
+			expectedHTTP: HTTPHealtcheck{Path: "/", Port: int32(4572)},
+		},
+		{
+			name:         "RegularCurlWithLocalhost",
+			test:         []string{"curl http://localhost:4572"},
+			expectedTest: HealtcheckTest{},
+			expectedHTTP: HTTPHealtcheck{Path: "/", Port: int32(4572)},
+		},
+		{
+			name:         "RegularCurlTrailingSlash",
+			test:         []string{"curl http://0.0.0.0:4572/"},
+			expectedTest: HealtcheckTest{},
+			expectedHTTP: HTTPHealtcheck{Path: "/", Port: int32(4572)},
+		},
+		{
+			name:         "MissingScheme",
+			test:         []string{"curl 0.0.0.0:4572/readiness"},
+			expectedTest: HealtcheckTest{},
+			expectedHTTP: HTTPHealtcheck{Path: "/readiness", Port: int32(4572)},
+		},
+		{
+			name:         "MissingSchemeWithF",
+			test:         []string{"curl -f localhost:8080/"},
+			expectedTest: HealtcheckTest{},
+			expectedHTTP: HTTPHealtcheck{Path: "/", Port: int32(8080)},
+		},
+		{
+			name:         "NoTest",
+			test:         []string{"curl --fail 0.0.0.0:8080"},
+			expectedTest: HealtcheckTest{},
+			expectedHTTP: HTTPHealtcheck{Path: "/", Port: int32(8080)},
+		},
+		{
+			name:         "Readiness",
+			test:         []string{"curl https://0.0.0.0:8080/readiness"},
+			expectedTest: HealtcheckTest{},
+			expectedHTTP: HTTPHealtcheck{Path: "/readiness", Port: int32(8080)},
+		},
+		{
+			name:         "RegularCurlWithPath",
+			test:         []string{"curl http://0.0.0.0:4572/a/path/exists"},
+			expectedTest: HealtcheckTest{},
+			expectedHTTP: HTTPHealtcheck{Path: "/a/path/exists", Port: int32(4572)},
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			check := HealthCheck{
+				Test: tc.test,
+			}
+			translateHealtcheckCurlToHTTP(&check)
+			if !reflect.DeepEqual(check.Test, tc.expectedTest) {
+				t.Fatalf("expected %v but got %v", tc.expectedTest, check.Test)
+			}
+			if !reflect.DeepEqual(check.Test, tc.expectedTest) {
+				t.Fatalf("expected %v but got %v", tc.expectedHTTP, check.HTTP)
+			}
+		})
+	}
+}
 
 func Test_DeployReplicasUnmarshalling(t *testing.T) {
 	tests := []struct {
@@ -324,7 +396,8 @@ func Test_HealthcheckUnmarshalling(t *testing.T) {
 			}
 			if !tt.expectedError {
 				if !reflect.DeepEqual(s.Services["app"].Healtcheck, tt.expected) {
-					t.Fatalf("Expected %v, but got %v", tt.expected, s.Services["app"].Healtcheck)
+					fmt.Printf("GOT: %+v", s.Services["app"].Healtcheck.HTTP)
+					t.Fatalf("Expected %+v, but got %+v", tt.expected, s.Services["app"].Healtcheck)
 				}
 			}
 

